feat(kubectl get, api): add nextScheduleTime field to CronJobStatus

choejwoo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
/kind api-change

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Adds a `nextScheduleTime` field to CronJobStatus to provide information about when the next job will be scheduled. This improves observability and helps users understand when their CronJobs will run next.

The field is calculated by the CronJob controller and provides users with visibility into the scheduling behavior of their CronJobs, making it easier to debug and monitor CronJob operations.

The next schedule time is displayed in the `NEXT SCHEDULE` column when using `kubectl get cronjob -o wide` format.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Closes #78564 

#### Special notes for your reviewer:

This PR adds a new field to the CronJobStatus API. The implementation:

1. Adds NextScheduleTime *metav1.Time field to CronJobStatus
2. Updates the CronJob controller to calculate and set the next schedule time
3. Updates printer code to display the field in wide output format
4. Includes comprehensive test coverage
5. Updates OpenAPI spec to include the new field

```yaml
apiVersion: batch/v1
kind: CronJob
metadata:
  name: hello
spec:
  schedule: "* * * * *"
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox:1.28
            imagePullPolicy: IfNotPresent
            command:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
      
```

```
✗ cluster/kubectl.sh get cronjob -o wide  
NAME    SCHEDULE    TIMEZONE   SUSPEND   ACTIVE   LAST SCHEDULE   NEXT SCHEDULE   AGE   CONTAINERS   IMAGES         SELECTOR
hello   * * * * *   <none>     False     1        32s             27s             70s   hello        busybox:1.28   <none>
```

After schedule field changed to `* * * * 1`

```
✗ cluster/kubectl.sh get cronjob -o wide  
NAME    SCHEDULE    TIMEZONE   SUSPEND   ACTIVE   LAST SCHEDULE   NEXT SCHEDULE   AGE    CONTAINERS   IMAGES         SELECTOR
hello   * * * * 1   <none>     False     2        31s             2d9h            2m9s   hello        busybox:1.28   <none>
```

Timezone Test
```
✗ cluster/kubectl.sh get cronjob -o wide
NAME                   SCHEDULE    TIMEZONE        SUSPEND   ACTIVE   LAST SCHEDULE   NEXT SCHEDULE   AGE   CONTAINERS   IMAGES         SELECTOR
test-london-timezone   * * * * 6   Europe/London   False     0        <none>          16h             8s    hello        busybox:1.28   <none>

✗ cluster/kubectl.sh get cronjob -o wide               
NAME                    SCHEDULE    TIMEZONE           SUSPEND   ACTIVE   LAST SCHEDULE   NEXT SCHEDULE   AGE   CONTAINERS   IMAGES         SELECTOR
test-newyork-timezone   * * * * 6   America/New_York   False     0        <none>          21h             27s   hello        busybox:1.28   <none>
```

The CronJobStatus now includes a `nextScheduleTime` field that displays when the next job will be scheduled. This provides users with better visibility into their CronJob scheduling behavior.


```yaml
status:
  active:
    ...
  lastScheduleTime: "2025-08-08T05:26:00Z"
  nextScheduleTime: "2025-08-10T15:00:00Z"
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add NextScheduleTime field to CronJobStatus

The CronJobStatus now includes a `nextScheduleTime` field that shows when the next job will be scheduled.

Users can see the next schedule time in:
- `kubectl get cronjob -o wide` output
- CronJob status API responses (`nextScheduleTime` new field added)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

choejwoo committed:
add next schedule time in cronjob status

Signed-off-by: choejwoo <jaewoo45@gmail.com>

choejwoo committed:
fix update status

Signed-off-by: choejwoo <jaewoo45@gmail.com>

choejwoo committed:
printer test code update

Signed-off-by: choejwoo <jaewoo45@gmail.com>

choejwoo committed:
Update OpenAPI spec for nextScheduleTime field

Signed-off-by: choejwoo <jaewoo45@gmail.com>

choejwoo committed:
Fix ineffectual assignment to nextScheduleTime in printers.go

Signed-off-by: choejwoo <jaewoo45@gmail.com>

choejwoo commented:
/label api-review

This PR adds only a status field (`nextScheduleTime`) to CronJobStatus to improve observability. 
No breaking changes or structural modifications to existing APIs.

liggitt commented:
Thanks for the PR. Before we dive into the details of API review, it would be helpful for sig-apps leads to ack the addition. https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#mechanics has an overview of this. API changes are connected to accepted designs (Kubernetes Enhancement Proposals, aka KEPs)

/assign @soltysh 

Once this has an associated design and is acked by sig-apps leads, we can pull it back into the API review pipeline

/remove-label api-review

yongruilin commented:
/remove-sig api-machinery

choejwoo commented:
@liggitt 
I was initially aiming to change only the `kubectl print` behavior and didn’t anticipate that this would involve adding an API field, so I hadn’t considered the KEP-related process. Thank you for sharing the guide. I’ll review the KEP procedure, and if it gets approved, I’ll request to re-add this to the API review pipeline. Thanks again!

https://github.com/kubernetes/enhancements/issues/5481

choejwoo committed:
fix: correct typos in cronjob and job controller comments

- Fix 'expexted' -> 'expected' in utils_test.go
- Fix 'rouge' -> 'rogue' in utils_test.go
- Fix 'expotential' -> 'exponential' in job_controller.go
- Fix 'MaxGeneneratedNameLength' -> 'MaxGeneratedNameLength' in indexed_job_utils_test.go

Signed-off-by: choejwoo <jaewoo45@gmail.com>

choejwoo committed:
staging: regenerate protobufs

Signed-off-by: choejwoo <jaewoo45@gmail.com>
